### DISCUSSION TITLE
Extract attachment from PDF file

### DISCUSSION THREAD

**Discussion by DianaC68** [qpdf/qpdf#545] [Q&A]
Hi, I would like to Extract attachment from PDF file. I use this command in VBA

Sub prova()
    path1 = "C:\Documenti\qpdf\qpdf\bin\qpdf"
    path2 = "C:\Users\Diana\"
    
    Shell (path1 & " " & "--password=" & Chr(34) & "MiaPassword" & Chr(34) & " " & Chr(34) & path2 & "Miofile.pdf" & Chr(34) & " " & "--list-attachments --verbose > " & Chr(34) & path2 & "elenco2.0.txt" & Chr(34))
End Sub
this command works correctly in MSDOS e Powershell, but doesn't it works in VBA. Can you help me? thanks

**Comment by jberkenbilt** [comment]
I don't know anything about VBA, but perhaps someone else may see this and provide an answer.

**Comment by DianaC68** [comment]
Thank you
Diana

Il Mar 13 Lug 2021, 16:47 Jay Berkenbilt ***@***.***> ha
scritto:

**Comment by dreua** [comment]
`" & Chr(34) & path2 & "Miofile.pdf" & Chr(34)` Maybe you just forgot one backslash here? I recommend saving the command in a variable and printing out so you can spot errors more easily. Also post any output/errors if possible!

**Comment by DianaC68** [comment]
Thanks very much biut slash is present in path2. I modified instruction in
order to write a shorter string.
Anyway, tomorrow I'll verify because now I am enable.
Diana

Il Mar 13 Lug 2021, 17:03 David Auer ***@***.***> ha scritto:

**Comment by DianaC68** [comment]
Goodmorning, I'll try to explain better. I'll write the original macro:

Sub prova()
path1 = "C:\Documenti\qpdf\qpdf\bin\qpdf.exe"
path2 = "C:\Users\S505160\Documents\Diana New\Sviluppi TEAM\CFO\Regolamento Titoli\Allegati\"

Shell (path1 & " " & "--password=" & Chr(34) & "MiaPassword" & Chr(34) & " " & Chr(34) & path2 & "[ENCRYPTED] attachments.pdf" & Chr(34) & " " & "--list-attachments --verbose > " & Chr(34) & path2 & "elenco2.0.txt" & Chr(34))
End Sub

This macro produces, with debug.print, the next string:
C:\Documenti\qpdf\qpdf\bin\qpdf.exe --password="Estero*01" "C:\Users\S505160\Documents\Diana New\Sviluppi TEAM\CFO\Regolamento Titoli\Allegati\[ENCRYPTED] attachments.pdf" --list-attachments --verbose > "C:\Users\S505160\Documents\Diana New\Sviluppi TEAM\CFO\Regolamento Titoli\Allegati\elenco2.0.txt"

If I'm going to run this string in PowerShell works good, but if I'm going to run with VBA, it doesn't work.
Someone can help me?
Thanks very much
Diana

**Reply by dreua** [reply]
Vba's Shell command runs on the regular command line (batch?), Not PowerShell. See the following SO answer for a possible solution: https://stackoverflow.com/a/32101992/4529404

**Comment by DianaC68** [comment]
Thank you for your assistance. I tried and thi si the result (parahaps I was wrong something?)

Sub prova()
Dim pscmd
path1 = "C:\Documenti\qpdf\qpdf\bin\qpdf.exe"
path2 = "C:\Users\S505160\Documents\Diana New\Sviluppi TEAM\CFO\Regolamento Titoli\Allegati\"

pscmd = "PowerShell -NoProfile -NoLogo -ExecutionPolicy Bypass -Command (" & path1 & " " & "--password=" & Chr(34) & "MyPassword" & Chr(34) & " " & Chr(34) & path2 & "[ENCRYPTED] attachments.pdf" & Chr(34) & " " & "--list-attachments " & Chr(34) & "--verbose" & Chr(34) & " > " & Chr(34) & path2 & "elenco2.0.txt" & Chr(34) & ")"


End Sub
With debug.print, I obtain this string:
PowerShell -NoProfile -NoLogo -ExecutionPolicy Bypass -Command (C:\Documenti\qpdf\qpdf\bin\qpdf.exe --password="MyPassword" "C:\Users\S505160\Documents\Diana New\Sviluppi TEAM\CFO\Regolamento Titoli\Allegati\[ENCRYPTED] attachments.pdf" --list-attachments "--verbose" > "C:\Users\S505160\Documents\Diana New\Sviluppi TEAM\CFO\Regolamento Titoli\Allegati\elenco2.0.txt")

Is it ok?

**Comment by DianaC68** [comment]
I'm sorry, but I forgotten thant this same command, i.e.
PowerShell -NoProfile -NoLogo -ExecutionPolicy Bypass -Command (C:\Documenti\qpdf\qpdf\bin\qpdf.exe --password="MyPassword" "C:\Users\S505160\Documents\Diana New\Sviluppi TEAM\CFO\Regolamento Titoli\Allegati[ENCRYPTED] attachments.pdf" --list-attachments "--verbose" > "C:\Users\S505160\Documents\Diana New\Sviluppi TEAM\CFO\Regolamento Titoli\Allegati\elenco2.0.txt")

it works well in MSDOS too. So it works wel in PowerShell and in MSDOS, but not in VBA

thanks

**Reply by dreua** [reply]
I'm sorry, can't help you any more with this because I don't actually use VBA myself, just looked for the obvious mistakes. At any rate, I believe this is more a VBA issue than a Qpdf issue. Maybe the redirection (">") doesn't work in VBA, maybe it can't run the external commands at all for security reasons. Hard to tell without a test environment and I can't spend my free time for setting one up.

**Reply by jberkenbilt** [reply]
Thank you @dreua for pitching in. It's nice that there are people out there who can jump in and help. @DianaC68 I also don't know anything about VBA and agree that this is not a qpdf issue. You might be better off on stackoverflow or some other forum where you can get more generalized help.

**Comment by DianaC68** [comment]
Thanks everyone.

Il Mer 14 Lug 2021, 17:41 Jay Berkenbilt ***@***.***> ha
scritto:
